lib/Platform/Intl/PlatformIntlApple.mm

```diff
@@ -116,26 +364,489 @@
   return x.compare(y);
 }
 
+// Implementation of
+// https://tc39.es/ecma402/#datetimeformat-objects
 struct DateTimeFormat::Impl {
-  std::u16string locale;
+  // Enum declaration for DateTimeFormat internal slots
+  // https://tc39.es/ecma402/#datetimeformat-objects
+  enum class FormatMatcher : size_t {Undefined = 0, Bestfit, Basic};
+  static std::vector<std::u16string> formatMatcherVector = {u"Undefined", u"Bestfit", u"Basic"};
+  enum class HourCycle : size_t {Undefined = 0, H11, H12, H23, H24};
+  static std::vector<std::u16string> hourCycleVector = {u"Undefined", u"H11", u"H12", u"H23", u"H24"};
+  enum class Weekday : size_t {Undefined = 0, Narrow, Short, Long};
+  static std::vector<std::u16string> weekdayVector = {"Undefined", u"Narrow", u"Short", u"Long"};
+  enum class Era : size_t {Undefined = 0, Narrow, Short, Long};
+  static std::vector<std::u16string> eraVector = {u"Undefined", u"Narrow", u"Short", u"Long"};
+  enum class Year : size_t {Undefined = 0, TwoDigit, Numeric};
+  static std::vector<std::u16string> yearVector = {u"Undefined", u"TwoDigit", u"Numeric"};
+  enum class Month : size_t {Undefined = 0, TwoDigit, Numeric, Narrow, Short, Long};
+  static std::vector<std::u16string> monthVector = {u"Undefined", u"TwoDigit", u"Numeric", u"Narrow", u"Short", u"Long"};
+  enum class Day : size_t {Undefined = 0, TwoDigit, Numeric};
+  static std::vector<std::u16string> dayVector = {u"Undefined", u"TwoDigit", u"Numeric"};
+  enum class DayPeriod : size_t {Undefined = 0, Narrow, Short, Long};
+  static std::vector<std::u16string> dayPeriodVector = {u"Undefined", u"Narrow", u"Short", u"Long"};
+  enum class Hour : size_t {Undefined = 0, TwoDigit, Numeric};
+  static std::vector<std::u16string> hourVector = {u"Undefined", u"TwoDigit", u"Numeric"};
+  enum class Minute : size_t {Undefined = 0, TwoDigit, Numeric};
+  static std::vector<std::u16string> minuteVector = {u"Undefined", u"TwoDigit", u"Numeric"};
+  enum class Second : size_t {Undefined = 0, TwoDigit, Numeric};
+  static std::vector<std::u16string> secondVector = {u"Undefined", u"TwoDigit", u"Numeric"};
+  enum class TimeZoneName : size_t {Undefined = 0, Short, Long};
+  static std::vector<std::u16string> timeZoneNameVector = {u"Undefined", u"Short", u"Long"};
+  enum class DateTimeStyle : size_t {Undefined = 0, Full, Long, Medium, Short};
+  static std::vector<std::u16string> dateTimeStyleVector = {u"Undefined", u"Full", u"Long", u"Medium", u"Short"};
+
+  static std::u16string formatMatcherString(const FormatMatcher &formatMatcher) {
+    return formatMatcherVector[static_cast<size_t>(formatMatcher)];
```

Comment:
The static lookup tables can be moved into the scope of the functions that use them. You should also make them `static` and `constexpr`. For example:
```suggestion
    static constexpr char16_t* formatMatchers[] = {u"Undefined", u"Bestfit", u"Basic"};
    return formatMatchers[static_cast<size_t>(formatMatcher)];
```